add support for TTL values with time-unit suffixes

abraxxa opened the pull request:
as supported by Bind9

no ISO 8601 support so far

olof commented:
Hey! Sorry for not getting back to you sooner, I hadn't noticed the pull request.

Really nice work, and seems to work well and matches bind9 implementation from my testing just now. A few notes;

* I'm thinking about backwards compatibility; it is a bind extension, do you know how well supported it is outside of bind? While I don't think this is a general concern since a well formed rfc1034 zone wouldn't have those unit suffixes, but for those few potential use cases that could be affected, would it be feasible to add an option to turn it off? (Default on seems fine.)
* The code implements combined units nicely, but the tests don't seem to show for it; how about modifying one of the tests to be `1h30m` instead? (And perhaps one like `1d12h30m` just to drive it home?) I also note that the comment saying that units should be treated case insensitive. Could you add a test for that as well?
* One test was lost, I understand why - your test is a superset of it; but the fact that the test suite verified that particular thing was lost as well; just keep the eval wrapper around your call, with the message specifying that we loaded a zonefile without specifying a TTL to the constructor (that's what I understand it was trying to test - but it's been a while since I looked at the code).
* Could you also split long lines, to less than 80 to follow the current style? (The longest lines are both regexps that use /x already anyways.) (And speaking of editorial nitpicking, add a trailing newline to `03-parser.t` :-))

Let me know if you have moved on, and I can take it from here, otherwise I look forward to do a second review, without as much delay on my part!

abraxxa commented:
Hi Olof!
Thanks for your super-nice response and the in-depth review!

>     * I'm thinking about backwards compatibility; it is a bind extension, do you know how well supported it is outside of bind? While I don't think this is a general concern since a well formed rfc1034 zone wouldn't have those unit suffixes, but for those few potential use cases that could be affected, would it be feasible to add an option to turn it off? (Default on seems fine.)
I haven't investigated that as we're only running bind9. All I found is from its docs at https://bind9.readthedocs.io/en/latest/reference.html#term-duration.
I think it's safe to always support these as such a formatted file failed to parse before.
 
>     * The code implements combined units nicely, but the tests don't seem to show for it; how about modifying one of the tests to be `1h30m` instead? (And perhaps one like `1d12h30m` just to drive it home?) I also note that the comment saying that units should be treated case insensitive. Could you add a test for that as well?
That's because I copied the test data from db.simple and just changed the format of the TTLs but not their values.
My idea was to run the complete test suite for both files so I wanted to keep the values identical.
But yes, this is important and I'll change the data and test. We can adapt the db.simple values later if we change the test to really run the same assertions for both files.

>     * One test was lost, I understand why - your test is a superset of it; but the fact that the test suite verified that particular thing was lost as well; just keep the eval wrapper around your call, with the message specifying that we loaded a zonefile without specifying a TTL to the constructor (that's what I understand it was trying to test - but it's been a while since I looked at the code).
I'm not sure I've removed this test intentionally so I've put it back.

>     * Could you also split long lines, to less than 80 to follow the current style? (The longest lines are both regexps that use /x already anyways.) (And speaking of editorial nitpicking, add a trailing newline to `03-parser.t` :-))
Yes, of course!

> Let me know if you have moved on, and I can take it from here, otherwise I look forward to do a second review, without as much delay on my part!
I've already force-pushed the changes, please review again.

abraxxa committed:
add support for time-unit suffixes

This change adds support for time unit suffixes (e.g. 10h, instead of
36000), as supported by Bind9, nsd and others. The supported time units
are:

 s second (*1)
 m minute (*60)
 h hour   (*3600)
 d day    (*86400)
 w week   (*604800)

The units can be combined, like `1h30m30s` (but not `1m30`: write
`1m30s`).

olof commented:
Hi, thanks a lot! I tested what nsd did with such values, and it supports them as well. I'm fine with not having the option to disable this.

Now this is probably not how you are supposed to use github, but i couldn't fine a nice way to link to an equivalent diff ([I tried](https://github.com/abraxxa/Parse-DNS-Zone/compare/master...olof:Parse-DNS-Zone:pr5)), but I have a number of mostly code style/naming changes, outlined in the diff below. The changes are also available as the [pr5 branch](https://github.com/olof/Parse-DNS-Zone/tree/pr5), which I believe is more useful to interact with. :) On that branch is also a small number of additional negative tests added, for things like `1m30` and invalid units.

I propose changing the name from `_parse_time_unit_with_suffix` to `_parse_time_unit`, as it handles both cases - and it makes for a shorter name at the same time. A couple of changes to the test data, so that it covers the "s" unit and upper case units as well. Please let me know what you think!

I have one question, and I haven't quite understood it yet: if an invalid unit is encountered, there is a die, supposed to trigger. I don't think this would be ideal, but on the other hand, I can't cause it to die in the tests. I think this case is already handled by the regexp, lacking those invalid units, and as such will never end up in `@parts`. Does this sound correct?

````diff
diff --git a/Changes b/Changes
index dd30c9d..99468af 100644
--- a/Changes
+++ b/Changes
@@ -4,8 +4,9 @@ Changes in Parse::DNS::Zone
 - Fix crash bug for get_rdata, when asking about non existing
   names/rdata in list context
 - Improvements to test suite
-- Add support for TTL values with time-unit suffixes
-  as supported by Bind9
+- Add support for TTL values with time-unit suffixes as
+  supported by Bind9, nsd et al (supported added by Alexander
+  Hartmaier)
 
 0.60 2016-11-11
 - Handle backslash escaped comment characters. Patch by Daniel
diff --git a/lib/Parse/DNS/Zone.pm b/lib/Parse/DNS/Zone.pm
index 28526e0..cfc1f29 100644
--- a/lib/Parse/DNS/Zone.pm
+++ b/lib/Parse/DNS/Zone.pm
@@ -53,6 +53,11 @@ RFC 1034:
 
 Parse::DNS::Zone does not support $GENERATE in this version.
 
+Additionally, use of time unit suffixes (e.g. using 1d instead of
+86400 or 1m30s instead of 90, etc), as supported by Bind9 and nsd et
+al is supported in TTLs and the time fields of a SOA record (as
+of version 0.70).
+
 =cut
 
 use 5.010;
@@ -408,16 +413,15 @@ sub _load_zonefile {
        return do { local $/; <$zonefh> }; # slurp
 }
 
-sub _parse_time_unit_with_suffix {
-       my $string = shift;
-
-       my $time_units_regex = qr/^
-                       (?: (?=\d+(?: w|d|h|m|s))
-                       (\d+w)? (\d+d)? (\d+h)? (\d+m)? (\d+s)? ) | (\d+)
-               $/xi;
-       my @parts = ($string =~ $time_units_regex);
+sub _parse_time_unit {
+       my @parts = shift =~ /^
+               (?:
+                       (?=\d+(?: w|d|h|m|s))
+                       (\d+w)? (\d+d)? (\d+h)? (\d+m)? (\d+s)?
+               ) | (\d+)
+       $/xi;
 
-       my %multi_for_unit = (
+       my %multiplier = (
                s => 1,
                m => 60,
                h => 3600,
@@ -428,17 +432,11 @@ sub _parse_time_unit_with_suffix {
        my $seconds = 0;
 
        for my $part (@parts) {
-               next
-                       if !defined $part;
-               my ($value, $unit) = ($part =~ /^(\d+)(\D+)?$/);
-               $unit = 's'
-                       if (!defined $unit || $unit eq '');
-               # from Bind reference manual:
-               # Both TTL-style and ISO 8601 duration formats are case-insensitive.
+               next unless $part;
+               my ($value, $unit) = $part =~ /^(\d+)(\D+)?$/;
+               $unit = 's' unless $unit;
                $unit = lc($unit);
-               die "unit $unit not supported"
-                       unless exists $multi_for_unit{$unit};
-               $seconds += $value * $multi_for_unit{$unit};
+               $seconds += $value * $multiplier{$unit};
        }
 
        return $seconds;
@@ -472,8 +470,9 @@ sub _parse_zone {
        my $time_units_regex = qr/
                (?:
                        (?=\d+(?: w|d|h|m|s))
-                       (?: \d+w)? (?: \d+d)? (?: \d+h)? (?: \d+m)? (?: \d+s)? ) | \d+
-               /ix;
+                       (?: \d+w)? (?: \d+d)? (?: \d+h)? (?: \d+m)? (?: \d+s)?
+               ) | \d+
+       /ix;
 
        my $zentry = qr/^
                (\S+)\s+ # name
@@ -539,7 +538,7 @@ sub _parse_zone {
 
                $origin = $1, next if(/^\$ORIGIN ([\w\-\.]+)\s*$/i);
                if(/^\$TTL (\S+)\s*$/i) {
-                       my $seconds = _parse_time_unit_with_suffix($1);
+                       my $seconds = _parse_time_unit($1);
                        $def_ttl = $seconds;
                        next;
                }
@@ -583,7 +582,7 @@ sub _parse_zone {
                        }
                }
 
-               $ttl = defined $ttl ?  _parse_time_unit_with_suffix($ttl) : $def_ttl;
+               $ttl = defined $ttl ?  _parse_time_unit($ttl) : $def_ttl;
                $class = defined $class ? $class : $def_class;
                $def_class = $class;
 
@@ -654,10 +653,10 @@ sub _parse_soa {
        $self->{soa}{mname}=$mname;
        $self->{soa}{rname}=$rname;
        $self->{soa}{serial}=$serial;
-       $self->{soa}{refresh}=_parse_time_unit_with_suffix($refresh);
-       $self->{soa}{retry}=_parse_time_unit_with_suffix($retry);
-       $self->{soa}{expire}=_parse_time_unit_with_suffix($expire);
-       $self->{soa}{minimum}=_parse_time_unit_with_suffix($minimum);
+       $self->{soa}{refresh}=_parse_time_unit($refresh);
+       $self->{soa}{retry}=_parse_time_unit($retry);
+       $self->{soa}{expire}=_parse_time_unit($expire);
+       $self->{soa}{minimum}=_parse_time_unit($minimum);
 }
 
 1;
diff --git a/t/03-parser.t b/t/03-parser.t
index 2fd5481..9228b61 100644
--- a/t/03-parser.t
+++ b/t/03-parser.t
@@ -336,7 +336,7 @@ is($zone->get_minimum, 3600, "SOA minimum");
 
 is(
        $zone->get_rdata(name=>'test-ttlunit', rr=>'A', field=>'ttl'),
-       '3600', 'Extr act ttl data from rr'
+       '3600', 'Extract ttl data from rr'
 );
 
-done_testing;
\ No newline at end of file
+done_testing;
diff --git a/t/data/db.units b/t/data/db.units
index 6240ab5..b872f21 100644
--- a/t/data/db.units
+++ b/t/data/db.units
@@ -3,7 +3,7 @@ $TTL   1h
 example.com.               IN   SOA       ns1.example.com. hostmaster.example.com. (
                                               1234567890   ; Serial
                                               1d12h30m     ; Refresh
-                                              1h      ; Retry
+                                              3600s      ; Retry
                                               14d      ; Expire
                                               1h  )   ; Negative Cache TTL
 
@@ -34,7 +34,7 @@ test-class                 IN   A         127.0.0.1
 test-ttl             400        A         127.0.0.1
 test-ttlclass        400   IN   A         127.0.0.1
 test-ttlclassr       IN   400   A         127.0.0.1
-test-ttlunit         1h         A         127.0.0.1
+test-ttlunit         1H         A         127.0.0.1
 
 test-origapp                    CNAME     test
````

abraxxa commented:
I'm used to the 'Files changed' tab of the PR for discussing code changes.

Renaming the function to `_parse_time_unit` is fine for me as well as the style changes.
The `die` is a safety net. I'm doing this of something similar for every hash lookup where the existence isn't checked somewhere else.

The only thing I don't like is the rename of `%multi_for_unit` to `%multiplier`. I'm following the naming convention of using plural for arrays and expressive names for hashes since quite some years. This makes it way easier to know what the expected key of the hash is, like the id or the name of something.

Should I take over your changes in my PR so it's easier the review them again and merge?

olof commented:
Hi, thanks! Yes, please do! And feel free to keep `multi_for_unit`, I don't feel strongly about that.

re: `die`, I see. That makes sense then!

abraxxa commented:
I've pushed the changes, please review (via the Files changed tab). Thanks!

olof commented on t/03-parser.t:
```diff
@@ -322,3 +322,21 @@ ok(
+	'3600', 'Extract ttl data from rr'
+);
+
+done_testing;
```
Your editor strips the final newline from this file for some reason.

abraxxa replied:
Yes, vscode does that and it annoyed me already. I'll investigate how to configure that and fix it with vim if you want me to.

olof commented on lib/Parse/DNS/Zone.pm:
```diff
@@ -408,6 +413,42 @@ sub _load_zonefile {
+
+	for my $part (@parts) {
+		next
+			unless $part;
```
I prefer these postfix `unless/if/for/etc` to be on a single line, when the expression is simple. *Especially* like in this case.

olof commented on lib/Parse/DNS/Zone.pm:
```diff
@@ -408,6 +413,42 @@ sub _load_zonefile {
+			unless $part;
+		my ($value, $unit) = $part =~ /^(\d+)(\D+)?$/;
+		$unit = 's'
+			unless $unit;
```
I prefer these postfix `unless/if/for/etc` to be on a single line, when the expression is simple.

(The `die` below is fine.)

olof commented on lib/Parse/DNS/Zone.pm:
```diff
@@ -408,6 +413,42 @@ sub _load_zonefile {
+		$unit = 's'
+			unless $unit;
+		# from Bind reference manual:
+		# Both TTL-style and ISO 8601 duration formats are case-insensitive.
```
I kind of think the comment is superfluous, especially in combination with tests covering this. At least remove mentions of ISO 8601, as we don't support it. (Fixing that would fix the line length as well. :).)

olof commented on lib/Parse/DNS/Zone.pm:
```diff
@@ -408,6 +413,42 @@ sub _load_zonefile {
+		# from Bind reference manual:
+		# Both TTL-style and ISO 8601 duration formats are case-insensitive.
+		$unit = lc($unit);
+		die "unit $unit not supported"
```
Perhaps change the error message to `BUG: Known unit is unhandled`, or what do you think? Makes it clear that it isn't supposed to happen.

(But I'm still pondering if it is a good idea to include it - this can never happen, as the regexp already validates the supported units. Only way for this to happen if the regexp and the hash diverges. But I can't imagine this set will ever change? And if so, having unit tests to cover it should be enough. But I won't force the issue.)

olof commented on lib/Parse/DNS/Zone.pm:
```diff
@@ -408,6 +413,42 @@ sub _load_zonefile {
+			(?: (?=\d+(?: w|d|h|m|s))
+			(\d+w)? (\d+d)? (\d+h)? (\d+m)? (\d+s)? ) | (\d+)
+		$/xi;
+	my @parts = ($string =~ $time_units_regex);
```
No need for `()`.

olof commented on lib/Parse/DNS/Zone.pm:
```diff
@@ -408,6 +413,42 @@ sub _load_zonefile {
+	my $time_units_regex = qr/^
+			(?: (?=\d+(?: w|d|h|m|s))
+			(\d+w)? (\d+d)? (\d+h)? (\d+m)? (\d+s)? ) | (\d+)
+		$/xi;
```
This is over-indented, the final `$/xi;` should line up with `my` (and the two lines above should as such be deintended one level. My patch also made a simplification where we avoid having `$string` and `$time_units_regexp`, and instead just applied the regexp directly to shift.

olof commented on lib/Parse/DNS/Zone.pm:
```diff
@@ -493,7 +544,11 @@ sub _parse_zone {
 		}
 
 		$origin = $1, next if(/^\$ORIGIN ([\w\-\.]+)\s*$/i);
-		$def_ttl = $1, next if(/^\$TTL (\d+)\s*$/i);
```
I would restore this line, and just changing it to do `_parse_time_unit($1)` instead. Any reason not to?

abraxxa replied:
I'm not used to this syntax and PBP doesn't recommend it: https://metacpan.org/pod/Perl::Critic::Policy::ValuesAndExpressions::ProhibitCommaSeparatedStatements

abraxxa replied:
I've changed the code to not have the $seconds variable, but it is different in that in only calls _parse_time_unit after the regex successfully matched, not always and skip afterwards.

olof reviewed (changes requested):
Please also update the commit message, along the lines of what I suggested in my pr5 branch.

Thanks for being patient! :) (The code style is already not super consistent, i admit.)

olof commented:
Sorry for not getting back to you quicker. Great work, this was merged in 90630b212ff656338e65df550cf46584f062701a, and included in release 0.70.
